python/mxnet/text/embedding.py

```diff
@@ -0,0 +1,722 @@
+            NDArray of shape=(len(tokens), self.vec_len).
+        """
+        to_reduce = False
+        if not isinstance(tokens, list):
+            tokens = [tokens]
+            to_reduce = True
+
+        indices = [self.token_to_idx[token] if token in self.token_to_idx
+                   else self.unknown_idx for token in tokens]
+
+        vecs = nd.Embedding(nd.array(indices), self.idx_to_vec,
+                            self.idx_to_vec.shape[0], self.idx_to_vec.shape[1])
+
+        return vecs[0] if to_reduce else vecs
+
+    def update_token_vectors(self, tokens, new_vectors):
+        """Updates embedding vectors for tokens.
+
+
+        Parameters
+        ----------
+        tokens : str or a list of strs.
+            A token or a list of tokens whose embedding vector are to be
+            updated.
+        new_vectors : mxnet.ndarray.NDArray
+            An NDArray to be assigned to the embedding vectors of `tokens`.
+            Its length must be equal to the number of `tokens` and its width
+            must be equal to the dimension of embeddings of the glossary. If
+            `tokens` is a singleton, it must be 1-D or 2-D. If `tokens` is a
+            list of multiple strings, it must be 2-D.
+        """
+
+        assert self.idx_to_vec is not None, \
+            'The property `idx_to_vec` has not been properly set.'
+
+        if not isinstance(tokens, list) or \
+                isinstance(tokens, list) and len(tokens) == 1:
+            assert isinstance(new_vectors, nd.NDArray) and \
+                len(new_vectors.shape) in {1, 2}, \
+                '`new_vectors` must be a 1-D or 2-D NDArray if `tokens` is a ' \
+                'singleton.'
+            if not isinstance(tokens, list):
+                tokens = [tokens]
+            if len(new_vectors.shape) == 1:
+                new_vectors = new_vectors.expand_dims(0)
+
+        else:
+            assert isinstance(tokens, list), \
+                '`tokens` must be a string or a list of strings'
+            assert isinstance(new_vectors, nd.NDArray) and \
+                len(new_vectors.shape) == 2, \
+                '`new_vectors` must be a 2-D NDArray if `tokens` is a list ' \
+                'of multiple strings.'
+        assert new_vectors.shape[0] == len(tokens), \
+            'The length of new_vectors must be equal to the number of tokens.'
+        assert new_vectors.shape[1] == self.vec_len, \
+            'The width of new_vectors must be equal to the dimension of ' \
+            'embeddings of the glossary.'
+
+        indices = []
+        for token in tokens:
+            if token in self.token_to_idx:
+                indices.append(self.token_to_idx[token])
+            else:
+                raise ValueError('Token %s is unknown. To update the embedding '
+                                 'vector for an unknown token, please specify '
+                                 'it explicitly as the `unknown_token` %s in '
```

Comment:
How can a user add a new token in embedding? Should there be a separate method for that?

Reply:
Thanks. Embedding class always loads from pre-trained files. User can add/set new tokens via glossary rather than via embedding.

Reply:
from mxnet import ndarray as nd
y = nd.array([[  0.,   1.,   2.,   3.,   4.],
     [  5.,   6.,   7.,   8.,   9.],
     [ 10.,  11.,  12.,  13.,  14.],
     [ 15.,  16.,  17.,  18.,  19.]])

x = nd.array([1, 3])

%timeit y[x]
%timeit nd.Embedding(x, y, 4, y.shape[1])

196 µs ± 11.8 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
83.5 µs ± 20.9 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Reply:
@reminisce

Reply:
`y[x]` is advanced indexing which is usually slower than other indexing operations due to two reasons.
1. Overhead of sanity checking and preprocessing advanced indices before calling backend ops.
2. In this case, the backend op used is `gather_nd`, which is expected for retrieving scattered elements from an ndarray. For a regular shape index like `[1, 3]` indexing the first dimension, operators such as `take` or `slice` are much more efficient than `gather_nd`.